### PR TITLE
Add a high-level overview of Robolectric's architecture

### PR DESCRIPTION
**Pull request by hoisie**
Currently there is no good place to get a high-level overview of Robolectric's complex architecture. Add an initial ARCHITECTURE.md file to describe it. This could help new contributors get a better understanding of all of the pieces of Robolectric.

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+(RAP)](https://github.com/robolectric/robolectric/blob/master/processor/src/main/java/org/robolectric/annotation/processing/RobolectricProcessor.java),
+and custom shadows that are commonly specified using `@Config(shadows = …)`. RAP
+is configured to process all of the shadow files that exist in Robolectric’s
+code, such as [framework
+shadows](https://github.com/robolectric/robolectric/tree/master/shadows/framework).
+It processes all of the source files and generates a source file that implements
+the
+[ShadowProvider](https://github.com/robolectric/robolectric/blob/master/shadowapi/src/main/java/org/robolectric/internal/ShadowProvider.java)
+interface. During runtime, Robolectric will use ServiceLoader to detect all
+ShadowProvider classes and load them as built-in shadows by default.
```
Comment (JuliaSullivanGoogle) [inline ARCHITECTURE.md:109]: Im a bit confused here. The sentence above seems to imply that there's one ShadowProvider that is built from every shadow, but this sentence seems to imply there are multiple ShadowProviders.

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+must be initialized. A Sandbox consists of some high-level structures that are
+necessary to run a Robolectric test. It primarily contains a
+[SandboxClassLoader](https://github.com/robolectric/robolectric/blob/master/sandbox/src/main/java/org/robolectric/internal/bytecode/SandboxClassLoader.java),
+which is a custom ClassLoader that is bound to a specific instrumented
+Android-all jar. Sandboxes also contain the ExecutorService that serves as the
+main thread (UI thread) as well as high-level instrumentation configuration. The
+SandboxClassLoader is installed as the default ClassLoader for the test method.
+When any Android class is requested, SandboxClassLoader will attempt to load the
+Android class from the instrumented Android-all Jar first. The primary goal of
+SandboxClassLoader is to ensure that classes from the stubs jar are not
```
Comment (JuliaSullivanGoogle) [inline ARCHITECTURE.md:124]: This is the first time the stubs jar is mentioned. Can we define what it is first and what it is for?

File: ARCHITECTURE.md (hunk 1 above)
Comment (hoisie) [reply ARCHITECTURE.md]: Fixed.

File: ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+native methods are invoked. To do this, Robolectric uses a system called Shadow
+classes.
+
+Shadow classes are Java classes that contain the replacement code of Android
+methods when they are invoked. Each shadow class is bound to specific Android
+classes and methods through annotations. There are currently hundreds of shadow
+classes that can be found
+[here](https://github.com/robolectric/robolectric/tree/master/shadows/framework/src/main/java/org/robolectric/shadows).
+
+Shadow classes may optionally contain public apis APIs can customize the
```
[the remark this answers is not in the document]
Comment (hoisie) [reply ARCHITECTURE.md]: Fixed.

File: ARCHITECTURE.md (hunk 2 above)
Comment (hoisie) [reply ARCHITECTURE.md]: The android.jar files that are contained in the Android SDK are also knows as the stubs jar. I use the terms interchangeably. I mention the android.jar files earlier in the document.

**Pull request merged**